File: kata-solutions/tell-dont-ask/tell-dont-ask-python/src/domain/Product.py
from decimal import Decimal, ROUND_HALF_UP

from src.domain.Category import Category
from dataclasses import dataclass


@dataclass(frozen = True)
class Product(object):
    _name: str
    _price: Decimal
    _category: Category

    def get_name(self):
       return self._name
    
    def get_price(self):
       return self._price

    def unitary_tax(self):
      return Decimal(self._price / Decimal(100) * self._category.get_tax_percentage()).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)
    
    def unitary_taxed_amount(self):
      return Decimal(self._price + self.unitary_tax()).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)

File: kata-solutions/tell-dont-ask/tell-dont-ask-java/app/src/main/java/domain/Product.java
package domain;

import java.math.BigDecimal;
import static java.math.BigDecimal.valueOf;
import static java.math.RoundingMode.HALF_UP;

public class Product {
    public final String name;
    public final BigDecimal price;
    public final Category category;

    public Product(String name, BigDecimal price, Category category) {
        this.name = name;
        this.price = price;
        this.category = category;
    }

    public BigDecimal unitaryTax() {
        return price.divide(valueOf(100)).multiply(category.taxPercentage).setScale(2, HALF_UP);
    }

    public BigDecimal unitaryTaxedAmount() {
        return price.add(unitaryTax()).setScale(2, HALF_UP);
    }

}